File: apps/frontend/src/stores/operator.ts
/**
 * Operator Store
 *
 * Zustand store for managing Kubernetes operator resources.
 */

import { create } from 'zustand';
import { persist } from 'zustand/middleware';
import api from '../services/api';
import {
  ResourceState,
  OperatorType,
  NotebookResource,
  TrainingJobResource,
  InferenceServiceResource,
  ClusterStatus,
  CreateNotebookRequest,
  CreateTrainingJobRequest,
  CreateInferenceServiceRequest,
} from '../types/operator';

interface OperatorState {
  // Resources
  notebooks: NotebookResource[];
  trainingJobs: TrainingJobResource[];
  inferenceServices: InferenceServiceResource[];
  clusterStatus: ClusterStatus | null;

  // UI State
  loading: boolean;
  error: string | null;

  // Actions
  fetchNotebooks: () => Promise<NotebookResource[]>;
  fetchTrainingJobs: () => Promise<TrainingJobResource[]>;
  fetchInferenceServices: () => Promise<InferenceServiceResource[]>;
  fetchClusterStatus: () => Promise<ClusterStatus>;

  // Notebook Actions
  createNotebook: (request: CreateNotebookRequest & { namespace?: string }) => Promise<NotebookResource>;
  getNotebook: (name: string, namespace?: string) => Promise<NotebookResource>;
  startNotebook: (name: string, namespace?: string) => Promise<void>;
  stopNotebook: (name: string, namespace?: string) => Promise<void>;
  deleteNotebook: (name: string, namespace?: string) => Promise<void>;

  // Training Job Actions
  createTrainingJob: (request: CreateTrainingJobRequest & { namespace?: string }) => Promise<TrainingJobResource>;
  getTrainingJob: (name: string, namespace?: string) => Promise<TrainingJobResource>;
  deleteTrainingJob: (name: string, namespace?: string) => Promise<void>;

  // Inference Service Actions
  createInferenceService: (request: CreateInferenceServiceRequest & { namespace?: string }) => Promise<InferenceServiceResource>;
  getInferenceService: (name: string, namespace?: string) => Promise<InferenceServiceResource>;
  scaleInferenceService: (name: string, replicas: number, namespace?: string) => Promise<void>;
  deleteInferenceService: (name: string, namespace?: string) => Promise<void>;

  // Cluster Actions
  installCrds: () => Promise<void>;

  // Utility
  clearError: () => void;
}

export const useOperatorStore = create<OperatorState>()(
  persist(
    (set, get) => ({
      // Initial state
      notebooks: [],
      trainingJobs: [],
      inferenceServices: [],
      clusterStatus: null,
      loading: false,
      error: null,

      // Fetch all notebooks
      fetchNotebooks: async () => {
        set({ loading: true, error: null });
        try {
          const response = await api.get('/operator/notebooks');
          const notebooks: NotebookResource[] = response.data.map((nb: any) => ({
            apiVersion: nb.apiVersion || 'studio.one-data.io/v1alpha1',
            kind: 'Notebook',
            metadata: {
              name: nb.name,
              namespace: nb.namespace || 'default',
              uid: nb.uid || `${nb.name}-${Date.now()}`,
              creationTimestamp: nb.createdAt,
            },
            spec: {
              image: nb.spec?.image || nb.image || 'jupyter/base-notebook:latest',
              cpu: nb.spec?.cpu || nb.cpu || '500m',
              memory: nb.spec?.memory || nb.memory || '1Gi',
              gpu: nb.spec?.gpu || nb.gpu,
              storage: nb.spec?.storage || nb.storage || '1Gi',
              ports: nb.spec?.ports || nb.ports || [8888],
              env: nb.spec?.env || nb.env || {},
              workspace: nb.spec?.workspace || nb.workspace,
              timeout: nb.spec?.timeout || nb.timeout,
              autoStop: nb.spec?.autoStop ?? nb.autoStop ?? true,
            },
            status: {
              phase: nb.phase || nb.status?.phase || ResourceState.UNKNOWN,
              conditions: nb.status?.conditions || [],
              observedGeneration: nb.status?.observedGeneration || 1,
              replicas: nb.replicas || nb.status?.replicas || 0,
              readyReplicas: nb.readyReplicas || nb.status?.readyReplicas || 0,
              availableReplicas: nb.status?.availableReplicas || 0,
              updatedReplicas: nb.status?.updatedReplicas || 0,
              jupyterURL: nb.jupyterURL || nb.status?.jupyterURL,
              tensorboardURL: nb.status?.tensorboardURL,
              createdAt: nb.createdAt,
            },
          }));
          set({ notebooks, loading: false });
          return notebooks;
        } catch (error: any) {
          set({ error: error.message, loading: false });
          throw error;
        }
      },

      // Fetch all training jobs
      fetchTrainingJobs: async () => {
        set({ loading: true, error: null });
        try {
          const response = await api.get('/operator/training-jobs');
          const jobs: TrainingJobResource[] = response.data.map((job: any) => ({
            apiVersion: job.apiVersion || 'studio.one-data.io/v1alpha1',
            kind: 'TrainingJob',
            metadata: {
              name: job.name,
              namespace: job.namespace || 'default',
              uid: job.uid || `${job.name}-${Date.now()}`,
              creationTimestamp: job.createdAt,
            },
            spec: {
              backend: job.spec?.backend || job.backend || 'pytorch',
              strategy: job.spec?.strategy || job.strategy || 'ddp',
              entryPoint: job.spec?.entryPoint || job.entry_point || 'train.py',
              entryPointArgs: job.spec?.entryPointArgs || job.entry_point_args || [],
              numNodes: job.spec?.numNodes || job.num_nodes || 1,
              numProcessesPerNode: job.spec?.numProcessesPerNode || job.num_processes_per_node || 1,
              modelUri: job.spec?.modelUri || job.model_uri || '',
              outputUri: job.spec?.outputUri || job.output_uri,
              tensorboard: job.spec?.tensorboard ?? job.tensorboard ?? false,
              dockerImage: job.spec?.dockerImage || job.docker_image,
              resources: job.spec?.resources || job.resources || {},
            },
            status: {
              phase: job.phase || job.status?.phase || ResourceState.UNKNOWN,
              conditions: job.status?.conditions || [],
              observedGeneration: job.status?.observedGeneration || 1,
              replicas: job.status?.replicas || 0,
              readyReplicas: job.status?.readyReplicas || 0,
              availableReplicas: job.status?.availableReplicas || 0,
              updatedReplicas: job.status?.updatedReplicas || 0,
              startedAt: job.startedAt || job.status?.startedAt,
              completedAt: job.status?.completedAt,
            },
          }));
          set({ trainingJobs: jobs, loading: false });
          return jobs;
        } catch (error: any) {
          set({ error: error.message, loading: false });
          throw error;
        }
      },

      // Fetch all inference services
      fetchInferenceServices: async () => {
        set({ loading: true, error: null });
        try {
          const response = await api.get('/operator/inference-services');
          const services: InferenceServiceResource[] = response.data.map((svc: any) => ({
            apiVersion: svc.apiVersion || 'studio.one-data.io/v1alpha1',
            kind: 'InferenceService',
            metadata: {
              name: svc.name,
              namespace: svc.namespace || 'default',
              uid: svc.uid || `${svc.name}-${Date.now()}`,
              creationTimestamp: svc.createdAt,
            },
            spec: {
              modelUri: svc.spec?.modelUri || svc.model_uri || '',
              predictorType: svc.spec?.predictorType || svc.predictor_type || 'custom',
              framework: svc.spec?.framework || svc.framework,
              replicas: svc.spec?.replicas || svc.replicas || 1,
              autoscalingEnabled: svc.spec?.autoscalingEnabled ?? svc.autoscaling_enabled ?? false,
              minReplicas: svc.spec?.minReplicas || svc.min_replicas || 1,
              maxReplicas: svc.spec?.maxReplicas || svc.max_replicas || 3,
              resources: svc.spec?.resources || svc.resources || {},
            },
            status: {
              phase: svc.phase || svc.status?.phase || ResourceState.UNKNOWN,
              conditions: svc.status?.conditions || [],
              observedGeneration: svc.status?.observedGeneration || 1,
              replicas: svc.replicas || svc.status?.replicas || 0,
              readyReplicas: svc.readyReplicas || svc.status?.readyReplicas || 0,
              availableReplicas: svc.status?.availableReplicas || 0,
              updatedReplicas: svc.status?.updatedReplicas || 0,
              serviceURL: svc.serviceURL || svc.status?.serviceURL,
            },
          }));
          set({ inferenceServices: services, loading: false });
          return services;
        } catch (error: any) {
          set({ error: error.message, loading: false });
          throw error;
        }
      },

      // Fetch cluster status
      fetchClusterStatus: async () => {
        try {
          const response = await api.get('/operator/cluster/status');
          const status: ClusterStatus = response.data;
          set({ clusterStatus: status });
          return status;
        } catch (error: any) {
          set({ error: error.message });
          throw error;
        }
      },

      // Create notebook
      createNotebook: async (request) => {
        set({ loading: true, error: null });
        try {
          const response = await api.post('/operator/notebooks', {
            name: request.name,
            image: request.image,
            cpu: request.cpu,
            memory: request.memory,
            gpu: request.gpu,
            storage: request.storage,
            ports: request.ports,
            env: request.env,
            workspace: request.workspace,
            timeout: request.timeout,
            autoStop: request.autoStop,
          }, {
            params: { namespace: request.namespace || 'default' },
          });
          set({ loading: false });
          await get().fetchNotebooks();
          return response.data;
        } catch (error: any) {
          set({ error: error.message, loading: false });
          throw error;
        }
      },

      // Get single notebook
      getNotebook: async (name, namespace = 'default') => {
        try {
          const response = await api.get(`/operator/notebooks/${name}`, {
            params: { namespace },
          });
          return response.data;
        } catch (error: any) {
          set({ error: error.message });
          throw error;
        }
      },

      // Start notebook
      startNotebook: async (name, namespace = 'default') => {
        try {
          await api.post(`/operator/notebooks/${name}/start`, null, {
            params: { namespace },
          });
          await get().fetchNotebooks();
        } catch (error: any) {
          set({ error: error.message });
          throw error;
        }
      },

      // Stop notebook
      stopNotebook: async (name, namespace = 'default') => {
        try {
          await api.post(`/operator/notebooks/${name}/stop`, null, {
            params: { namespace },
          });
          await get().fetchNotebooks();
        } catch (error: any) {
          set({ error: error.message });
          throw error;
        }
      },

      // Delete notebook
      deleteNotebook: async (name, namespace = 'default') => {
        try {
          await api.delete(`/operator/notebooks/${name}`, {
            params: { namespace },
          });
          await get().fetchNotebooks();
        } catch (error: any) {
          set({ error: error.message });
          throw error;
        }
      },

      // Create training job
      createTrainingJob: async (request) => {
        set({ loading: true, error: null });
        try {
          const response = await api.post('/operator/training-jobs', {
            name: request.name,
            backend: request.backend,
            strategy: request.strategy,
            entry_point: request.entryPoint,
            entry_point_args: request.entryPointArgs,
            num_nodes: request.numNodes,
            num_processes_per_node: request.numProcessesPerNode,
            model_uri: request.modelUri,
            output_uri: request.outputUri,
            tensorboard: request.tensorboard,
            docker_image: request.dockerImage,
            resources: request.resources,
          }, {
            params: { namespace: request.namespace || 'default' },
          });
          set({ loading: false });
          await get().fetchTrainingJobs();
          return response.data;
        } catch (error: any) {
          set({ error: error.message, loading: false });
          throw error;
        }
      },

      // Get single training job
      getTrainingJob: async (name, namespace = 'default') => {
        try {
          const response = await api.get(`/operator/training-jobs/${name}`, {
            params: { namespace },
          });
          return response.data;
        } catch (error: any) {
          set({ error: error.message });
          throw error;
        }
      },

      // Delete training job
      deleteTrainingJob: async (name, namespace = 'default') => {
        try {
          await api.delete(`/operator/training-jobs/${name}`, {
            params: { namespace },
          });
          await get().fetchTrainingJobs();
        } catch (error: any) {
          set({ error: error.message });
          throw error;
        }
      },

      // Create inference service
      createInferenceService: async (request) => {
        set({ loading: true, error: null });
        try {
          const response = await api.post('/operator/inference-services', {
            name: request.name,
            model_uri: request.modelUri,
            predictor_type: request.predictorType,
            framework: request.framework,
            replicas: request.replicas,
            autoscaling_enabled: request.autoscalingEnabled,
            min_replicas: request.minReplicas,
            max_replicas: request.maxReplicas,
            resources: request.resources,
          }, {
            params: { namespace: request.namespace || 'default' },
          });
          set({ loading: false });
          await get().fetchInferenceServices();
          return response.data;
        } catch (error: any) {
          set({ error: error.message, loading: false });
          throw error;
        }
      },

      // Get single inference service
      getInferenceService: async (name, namespace = 'default') => {
        try {
          const response = await api.get(`/operator/inference-services/${name}`, {
            params: { namespace },
          });
          return response.data;
        } catch (error: any) {
          set({ error: error.message });
          throw error;
        }
      },

      // Scale inference service
      scaleInferenceService: async (name, replicas, namespace = 'default') => {
        try {
          await api.put(`/operator/inference-services/${name}/scale`, {
            replicas,
          }, {
            params: { namespace },
          });
          await get().fetchInferenceServices();
        } catch (error: any) {
          set({ error: error.message });
          throw error;
        }
      },

      // Delete inference service
      deleteInferenceService: async (name, namespace = 'default') => {
        try {
          await api.delete(`/operator/inference-services/${name}`, {
            params: { namespace },
          });
          await get().fetchInferenceServices();
        } catch (error: any) {
          set({ error: error.message });
          throw error;
        }
      },

      // Install CRDs
      installCrds: async () => {
        set({ loading: true, error: null });
        try {
          await api.post('/operator/cluster/install-crds');
          set({ loading: false });
        } catch (error: any) {
          set({ error: error.message, loading: false });
          throw error;
        }
      },

      // Clear error
      clearError: () => set({ error: null }),
    }),
    {
      name: 'operator-storage',
      partialize: (state) => ({
        clusterStatus: state.clusterStatus,
      }),
    }
  )
);

// Selectors
export const selectRunningNotebooks = (state: OperatorState) =>
  state.notebooks.filter((n) => n.status.phase === ResourceState.RUNNING);

export const selectRunningTrainingJobs = (state: OperatorState) =>
  state.trainingJobs.filter((j) => j.status.phase === ResourceState.RUNNING);

export const selectRunningInferenceServices = (state: OperatorState) =>
  state.inferenceServices.filter((s) => s.status.phase === ResourceState.RUNNING);

export const selectFailedResources = (state: OperatorState) => ({
  notebooks: state.notebooks.filter((n) => n.status.phase === ResourceState.FAILED),
  trainingJobs: state.trainingJobs.filter((j) => j.status.phase === ResourceState.FAILED),
  inferenceServices: state.inferenceServices.filter((s) => s.status.phase === ResourceState.FAILED),
});

File: apps/backend/app/api/v1/operator.py
"""
Kubernetes Operator API Endpoints

Provides REST API for managing custom resources through operators.
"""

import logging
from typing import List, Optional, Dict, Any
from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel, Field

from app.api.deps import get_current_user
from app.models.user import User
from app.services.operator import (
    ResourceState,
    get_operator_manager,
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/operator", tags=["Kubernetes Operator"])


# ============================================================================
# Request/Response Schemas
# ============================================================================


class CreateNotebookRequest(BaseModel):
    """Request to create notebook"""
    name: str = Field(..., description="Notebook name")
    image: str = Field(..., description="Container image")
    cpu: str = "500m"
    memory: str = "1Gi"
    gpu: Optional[int] = None
    storage: str = "1Gi"
    ports: List[int] = Field(default_factory=lambda: [8888])
    env: Dict[str, str] = Field(default_factory=dict)
    workspace: Optional[str] = None
    timeout: Optional[int] = None
    auto_stop: bool = True


class CreateTrainingJobRequest(BaseModel):
    """Request to create training job"""
    name: str
    backend: str  # pytorch, tensorflow
    strategy: str  # ddp, mirrored, etc.
    entry_point: str
    entry_point_args: List[str] = Field(default_factory=list)
    num_nodes: int = 1
    num_processes_per_node: int = 1
    model_uri: str
    output_uri: Optional[str] = None
    tensorboard: bool = False
    docker_image: Optional[str] = None
    resources: Dict[str, Any] = Field(default_factory=dict)


class CreateInferenceServiceRequest(BaseModel):
    """Request to create inference service"""
    name: str
    model_uri: str
    predictor_type: str = "custom"
    framework: Optional[str] = None
    replicas: int = 1
    autoscaling_enabled: bool = False
    min_replicas: int = 1
    max_replicas: int = 3
    resources: Dict[str, Any] = Field(default_factory=dict)


class ScaleRequest(BaseModel):
    """Request to scale resource"""
    replicas: int = Field(..., ge=0, le=100)


# ============================================================================
# Notebook Endpoints
# ============================================================================


@router.post("/notebooks", response_model=Dict[str, Any])
async def create_notebook(
    request: CreateNotebookRequest,
    namespace: str = "default",
    current_user: User = Depends(get_current_user),
):
    """Create a notebook server"""
    try:
        manager = get_operator_manager(namespace)

        spec = {
            "image": request.image,
            "cpu": request.cpu,
            "memory": request.memory,
            "ports": request.ports,
            "env": request.env,
            "auto_stop": request.auto_stop,
        }

        if request.gpu:
            spec["gpu"] = request.gpu
        if request.storage:
            spec["storage"] = request.storage
        if request.workspace:
            spec["workspace"] = request.workspace
        if request.timeout:
            spec["timeout"] = request.timeout

        resource = await manager.create_notebook(
            request.name,
            spec,
            current_user.id,
        )

        return {
            "name": resource["metadata"]["name"],
            "namespace": resource["metadata"]["namespace"],
            "phase": resource["status"]["phase"],
            "message": "Notebook created successfully",
        }
    except Exception as e:
        logger.error(f"Failed to create notebook: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )


@router.get("/notebooks", response_model=List[Dict[str, Any]])
async def list_notebooks(
    namespace: str = "default",
    current_user: User = Depends(get_current_user),
):
    """List all notebooks"""
    try:
        manager = get_operator_manager(namespace)
        notebooks = manager.list_notebooks()

        return [
            {
                "name": nb["metadata"]["name"],
                "namespace": nb["metadata"]["namespace"],
                "phase": nb["status"]["phase"],
                "readyReplicas": nb["status"].get("readyReplicas", 0),
                "replicas": nb["status"].get("replicas", 0),
                "jupyterURL": nb["status"].get("jupyterURL"),
                "createdAt": nb["metadata"].get("creationTimestamp"),
            }
            for nb in notebooks
        ]
    except Exception as e:
        logger.error(f"Failed to list notebooks: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )


@router.get("/notebooks/{name}")
async def get_notebook(
    name: str,
    namespace: str = "default",
    current_user: User = Depends(get_current_user),
):
    """Get notebook details"""
    try:
        manager = get_operator_manager(namespace)
        notebooks = manager.list_notebooks()

        for nb in notebooks:
            if nb["metadata"]["name"] == name:
                return {
                    "name": nb["metadata"]["name"],
                    "namespace": nb["metadata"]["namespace"],
                    "spec": nb["spec"],
                    "status": nb["status"],
                }

        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Notebook {name} not found",
        )
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to get notebook: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )


@router.delete("/notebooks/{name}")
async def delete_notebook(
    name: str,
    namespace: str = "default",
    current_user: User = Depends(get_current_user),
):
    """Delete a notebook"""
    try:
        manager = get_operator_manager(namespace)
        success = await manager.delete_notebook(name)

        if not success:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Notebook {name} not found",
            )

        return {"message": f"Notebook {name} deleted successfully"}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to delete notebook: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )


@router.post("/notebooks/{name}/start")
async def start_notebook(
    name: str,
    namespace: str = "default",
    current_user: User = Depends(get_current_user),
):
    """Start a notebook"""
    try:
        manager = get_operator_manager(namespace)
        # In a real implementation, this would start the notebook pods
        return {"message": f"Notebook {name} started"}
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )


@router.post("/notebooks/{name}/stop")
async def stop_notebook(
    name: str,
    namespace: str = "default",
    current_user: User = Depends(get_current_user),
):
    """Stop a notebook"""
    try:
        manager = get_operator_manager(namespace)
        # In a real implementation, this would stop the notebook pods
        return {"message": f"Notebook {name} stopped"}
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )


# ============================================================================
# Training Job Endpoints
# ============================================================================


@router.post("/training-jobs", response_model=Dict[str, Any])
async def create_training_job(
    request: CreateTrainingJobRequest,
    namespace: str = "default",
    current_user: User = Depends(get_current_user),
):
    """Create a training job"""
    try:
        manager = get_operator_manager(namespace)

        spec = {
            "backend": request.backend,
            "strategy": request.strategy,
            "entry_point": request.entry_point,
            "entry_point_args": request.entry_point_args,
            "num_nodes": request.num_nodes,
            "num_processes_per_node": request.num_processes_per_node,
            "model_uri": request.model_uri,
            "tensorboard": request.tensorboard,
        }

        if request.output_uri:
            spec["output_uri"] = request.output_uri
        if request.docker_image:
            spec["docker_image"] = request.docker_image
        if request.resources:
            spec["resources"] = request.resources

        resource = await manager.create_training_job(
            request.name,
            spec,
            current_user.id,
        )

        return {
            "name": resource["metadata"]["name"],
            "namespace": resource["metadata"]["namespace"],
            "phase": resource["status"]["phase"],
            "message": "Training job created successfully",
        }
    except Exception as e:
        logger.error(f"Failed to create training job: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )


@router.get("/training-jobs", response_model=List[Dict[str, Any]])
async def list_training_jobs(
    namespace: str = "default",
    current_user: User = Depends(get_current_user),
):
    """List all training jobs"""
    try:
        manager = get_operator_manager(namespace)
        jobs = manager.list_training_jobs()

        return [
            {
                "name": job["metadata"]["name"],
                "namespace": job["metadata"]["namespace"],
                "spec": job["spec"],
                "phase": job["status"]["phase"],
                "startedAt": job["status"].get("startedAt"),
            }
            for job in jobs
        ]
    except Exception as e:
        logger.error(f"Failed to list training jobs: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )


@router.delete("/training-jobs/{name}")
async def delete_training_job(
    name: str,
    namespace: str = "default",
    current_user: User = Depends(get_current_user),
):
    """Delete a training job"""
    try:
        manager = get_operator_manager(namespace)
        success = await manager.delete_training_job(name)

        if not success:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Training job {name} not found",
            )

        return {"message": f"Training job {name} deleted successfully"}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to delete training job: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )


# ============================================================================
# Inference Service Endpoints
# ============================================================================


@router.post("/inference-services", response_model=Dict[str, Any])
async def create_inference_service(
    request: CreateInferenceServiceRequest,
    namespace: str = "default",
    current_user: User = Depends(get_current_user),
):
    """Create an inference service"""
    try:
        manager = get_operator_manager(namespace)

        spec = {
            "model_uri": request.model_uri,
            "predictor_type": request.predictor_type,
            "replicas": request.replicas,
            "autoscaling_enabled": request.autoscaling_enabled,
        }

        if request.framework:
            spec["framework"] = request.framework
        if request.resources:
            spec["resources"] = request.resources

        resource = await manager.create_inference_service(
            request.name,
            spec,
            current_user.id,
        )

        return {
            "name": resource["metadata"]["name"],
            "namespace": resource["metadata"]["namespace"],
            "phase": resource["status"]["phase"],
            "message": "Inference service created successfully",
        }
    except Exception as e:
        logger.error(f"Failed to create inference service: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )


@router.get("/inference-services", response_model=List[Dict[str, Any]])
async def list_inference_services(
    namespace: str = "default",
    current_user: User = Depends(get_current_user),
):
    """List all inference services"""
    try:
        manager = get_operator_manager(namespace)
        services = manager.list_inference_services()

        return [
            {
                "name": svc["metadata"]["name"],
                "namespace": svc["metadata"]["namespace"],
                "spec": svc["spec"],
                "phase": svc["status"]["phase"],
                "replicas": svc["status"].get("replicas", 0),
                "readyReplicas": svc["status"].get("readyReplicas", 0),
                "serviceURL": svc["status"].get("serviceURL"),
            }
            for svc in services
        ]
    except Exception as e:
        logger.error(f"Failed to list inference services: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )


@router.delete("/inference-services/{name}")
async def delete_inference_service(
    name: str,
    namespace: str = "default",
    current_user: User = Depends(get_current_user),
):
    """Delete an inference service"""
    try:
        manager = get_operator_manager(namespace)
        success = await manager.delete_inference_service(name)

        if not success:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Inference service {name} not found",
            )

        return {"message": f"Inference service {name} deleted successfully"}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to delete inference service: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )


@router.put("/inference-services/{name}/scale")
async def scale_inference_service(
    name: str,
    request: ScaleRequest,
    namespace: str = "default",
    current_user: User = Depends(get_current_user),
):
    """Scale an inference service"""
    try:
        manager = get_operator_manager(namespace)
        services = manager.list_inference_services()

        for svc in services:
            if svc["metadata"]["name"] == name:
                # Update replicas
                if "status" not in svc:
                    svc["status"] = {}
                svc["status"]["replicas"] = request.replicas
                svc["status"]["readyReplicas"] = request.replicas

                # Update spec
                if "spec" not in svc:
                    svc["spec"] = {}
                svc["spec"]["replicas"] = request.replicas

                return {
                    "name": name,
                    "replicas": request.replicas,
                    "message": "Service scaled successfully",
                }

        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Inference service {name} not found",
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )


# ============================================================================
# Cluster Endpoints
# ============================================================================


@router.get("/cluster/status")
async def get_cluster_status(
    current_user: User = Depends(get_current_user),
):
    """Get operator cluster status"""
    try:
        return {
            "operators": {
                "notebook": {"running": True, "version": "v1alpha1"},
                "training": {"running": True, "version": "v1alpha1"},
                "inference": {"running": True, "version": "v1alpha1"},
            },
            "crds": {
                "notebooks": {"installed": True},
                "trainingjobs": {"installed": True},
                "inferenceservices": {"installed": True},
            },
        }
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )


@router.post("/cluster/install-crds")
async def install_crds(
    current_user: User = Depends(get_current_user),
):
    """Install CRDs to the cluster"""
    try:
        manager = get_operator_manager()
        results = await manager.install_crds()

        return {
            "message": "CRDs installed successfully",
            "results": results,
        }
    except Exception as e:
        logger.error(f"Failed to install CRDs: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )
